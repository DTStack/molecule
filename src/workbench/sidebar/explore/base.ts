import {
    getBEMElement,
    getBEMModifier,
    prefixClaName,
} from 'mo/common/className';
import {
    ID_ACTIVITY_BAR,
    ID_SIDE_BAR,
    ID_EXPLORER,
    ID_FOLDER_TREE,
    ID_EDITOR_TREE,
} from 'mo/common/id';

const defaultClassName = prefixClaName(ID_SIDE_BAR);
const defaultExplorerClassName = prefixClaName(ID_EXPLORER, defaultClassName);
const activityBarItemFloatClassName = getBEMModifier(
    getBEMElement(defaultExplorerClassName, ID_ACTIVITY_BAR),
    'float'
);

const folderTreeClassName = prefixClaName(ID_FOLDER_TREE);
const folderTreeInputClassName = getBEMModifier(folderTreeClassName, 'input');
const folderTreeEditClassName = getBEMModifier(folderTreeClassName, 'editable');

const editorTreeClassName = prefixClaName(ID_EDITOR_TREE);
const editorTreeItemClassName = getBEMElement(editorTreeClassName, 'item');
const editorTreeGroupClassName = getBEMElement(editorTreeClassName, 'group');
const editorTreeFileNameClassName = getBEMElement(
    editorTreeItemClassName,
    'fileName'
);
const editorTreeFilePathClassName = getBEMElement(
    editorTreeItemClassName,
    'filePath'
);
const editorTreeActiveItemClassName = getBEMModifier(
    editorTreeItemClassName,
    'active'
);
const editorTreeCloseIconClassName = getBEMElement(
    editorTreeClassName,
    'close'
);
const editorTreeFileIconClassName = getBEMElement(editorTreeClassName, 'file');

export {
    defaultExplorerClassName,
    activityBarItemFloatClassName,
    folderTreeClassName,
    folderTreeInputClassName,
    folderTreeEditClassName,
    editorTreeClassName,
    editorTreeItemClassName,
    editorTreeGroupClassName,
    editorTreeFileNameClassName,
    editorTreeFilePathClassName,
    editorTreeActiveItemClassName,
    editorTreeCloseIconClassName,
    editorTreeFileIconClassName,
};
