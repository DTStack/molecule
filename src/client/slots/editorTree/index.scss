@import '../../classNames/style.scss';

$editorTree: prefix('editorTree');
$group: BEMElement($editorTree, 'group');
$item: BEMElement($editorTree, 'item');
$toolbar: BEMElement($editorTree, 'toolbar');
$close: BEMElement($item, 'close');
$file: BEMElement($item, 'file');
$name: BEMElement($item, 'name');
$path: BEMElement($item, 'path');

$active: BEMModifier($item, 'active');

.#{$editorTree} {
    font-size: 13px;
    max-height: 220px;

    .#{$item} {
        // keep same with tree node
        align-items: center;
        outline: 1px solid transparent;
        outline-offset: -1px;
        cursor: pointer;
        display: flex;
        height: 22px;
        line-height: 22px;
        list-style: none;
        margin: 0;
        padding-left: 16px;
        user-select: none;
        white-space: nowrap;

        &.#{$active} {
            background-color: var(--list-inactiveSelectionBackground);

            .#{$close} {
                color: var(--icon-foreground);
                opacity: 1;
            }
        }

        &:hover {
            &:not(.#{$active}) {
                background-color: var(--list-hoverBackground);
            }

            .#{$close} {
                opacity: 1;
            }
        }

        &:focus {
            background: var(--list-activeSelectionBackground);
            outline-color: var(--list-focusOutline);
            color: #fff;
        }

        .#{$name} {
            flex-basis: auto;
            margin-left: 5px;
            overflow: hidden;
            text-overflow: ellipsis;
            white-space: pre;
        }

        .#{$path} {
            flex: 1;
            font-size: 0.9em;
            margin-left: 5px;
            opacity: 0.8;
            overflow: hidden;
            padding-top: 2px;
            text-overflow: ellipsis;
            white-space: pre;
        }
    }

    .#{$group} {
        font-size: 11px;
        height: 22px;
        line-height: 22px;
        padding-left: 16px;
        user-select: none;

        &:hover {
            background-color: var(--list-hoverBackground);

            .#{$toolbar} {
                opacity: 1;
            }
        }
    }

    .#{$toolbar} {
        opacity: 0;
    }

    .#{$close} {
        margin-right: 2px;
        opacity: 0;
    }
}

:export {
    editorTree: $editorTree;
    group: $group;
    item: $item;
    toolbar: $toolbar;
    close: $close;
    file: $file;
    name: $name;
    path: $path;

    active: $active;
}
