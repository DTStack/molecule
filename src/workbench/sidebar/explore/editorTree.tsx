import React, { useLayoutEffect, useRef } from 'react';
import { IEditorTreeController } from 'mo/controller';
import { IEditor, IEditorGroup } from 'mo/model';
import {
    IActionBarItemProps,
    Icon,
    IMenuItemProps,
    ITabProps,
    Menu,
    Scrollable,
    Toolbar,
    useContextView,
} from 'mo/components';
import {
    editorTreeActiveItemClassName,
    editorTreeClassName,
    editorTreeCloseIconClassName,
    editorTreeFileIconClassName,
    editorTreeFileNameClassName,
    editorTreeFilePathClassName,
    editorTreeGroupClassName,
    editorTreeItemClassName,
} from './base';
import { classNames } from 'mo/common/className';
import { getEventPosition } from 'mo/common/dom';
import { localize } from 'mo/i18n/localize';
import {
    ICollapseItem,
    HEADER_HEIGTH,
    MAX_GROW_HEIGHT,
} from 'mo/components/collapse';
import Scrollbar from 'react-scrollbars-custom';
import { constants } from 'mo/services/builtinService/const';
import type { UniqueId } from 'mo/common/types';

// override onContextMenu
type UnionEditor = Omit<
    IEditor & IEditorTreeController,
    'onContextMenu' | 'initView'
>;
export interface IOpenEditProps extends UnionEditor {
    /**
     * Group Header toolbar
     */
    groupToolbar?: IActionBarItemProps<IEditorGroup>[];
    /**
     * Item context menus
     */
    contextMenu?: IMenuItemProps[];
    /**
     * Group Header context menus
     * It'll use the value of contextMenu if specify contextMenu but not specify headerContextMenu
     */
    headerContextMenu?: IMenuItemProps[];
    onContextMenu?: (
        menu: IMenuItemProps,
        groupId: UniqueId,
        file?: ITabProps
    ) => void;
    panel: ICollapseItem;
}

const EditorTree = (props: IOpenEditProps) => {
    const {
        current,
        groups,
        groupToolbar,
        contextMenu = [],
        headerContextMenu,
        panel,
        onSelect,
        onSaveGroup,
        onContextMenu,
        onCloseGroup,
        onClose,
        onToolbarClick,
    } = props;

    const wrapper = useRef<HTMLDivElement>(null);
    const scrollable = useRef<Scrollbar>(null);

    // scroll into view
    useLayoutEffect(() => {
        const scrollHeight = scrollable.current?.scrollHeight || 0;
        if (scrollHeight > MAX_GROW_HEIGHT - HEADER_HEIGTH) {
            const activeItem = wrapper.current?.querySelector<HTMLDivElement>(
                `.${editorTreeActiveItemClassName}`
            );
            if (activeItem) {
                const top = activeItem.offsetTop;
                scrollable.current?.scrollTo(0, top);
            }
        }
    }, [current?.id && current.tab?.id]);

    if (!groups || !groups.length) return null;

    const contextView = useContextView();

    const handleCloseClick = (group: IEditorGroup, file: ITabProps) => {
        onClose?.(file.id!, group.id!);
    };

    const handleItemClick = (group: IEditorGroup, file: ITabProps) => {
        if (group.id !== current?.id || file.id !== current?.tab?.id) {
            onSelect?.(file.id!, group.id!);
        }
    };

    const handleOnMenuClick = (
        menu: IMenuItemProps,
        group: IEditorGroup,
        file?: ITabProps
    ) => {
        contextView.hide();
        onContextMenu?.(menu, group.id!, file);
    };

    const handleRightClick = (
        e: React.MouseEvent<HTMLDivElement, MouseEvent>,
        group: IEditorGroup,
        file: ITabProps
    ) => {
        e.preventDefault();
        contextView.show(getEventPosition(e), () => (
            <Menu
                role="menu"
                onClick={(_, item) => handleOnMenuClick(item!, group, file)}
                data={contextMenu}
            />
        ));
    };

    const handleHeaderRightClick = (
        e: React.MouseEvent<HTMLDivElement, MouseEvent>,
        group: IEditorGroup
    ) => {
        e.preventDefault();
        const groupHeaderContext = headerContextMenu || contextMenu;
        contextView.show(getEventPosition(e), () => (
            <Menu
                role="menu"
                onClick={(_, item) => handleOnMenuClick(item!, group)}
                data={groupHeaderContext}
            />
        ));
    };

    // click group title will open the first file in this group
    const handleGroupClick = (e, group: IEditorGroup) => {
        const { target } = e;
        const firstFile = group.data?.[0];
        if (target.nextElementSibling && firstFile) {
            onSelect?.(firstFile.id!, group.id!);
            target.nextElementSibling.focus();
        }
    };

    const handleToolBarClick = (
        e: React.MouseEvent<Element, MouseEvent>,
        item: IActionBarItemProps,
        group: IEditorGroup
    ) => {
        e.stopPropagation();
        switch (item.id) {
            case constants.EXPLORER_TOGGLE_CLOSE_GROUP_EDITORS:
                onCloseGroup?.(group.id!);
                break;
            case constants.EXPLORER_TOGGLE_SAVE_GROUP:
                onSaveGroup?.(group.id!);
                break;
            default:
                // default behavior
                onToolbarClick?.(item, group.id!);
                break;
        }
    };

    return (
        <Scrollable noScrollX isShowShadow ref={scrollable}>
            <div
                className={editorTreeClassName}
                ref={wrapper}
                data-content={panel.id}
            >
                {groups.map((group, index) => {
                    return (
                        <React.Fragment key={index}>
                            {groups.length !== 1 && (
                                <div
                                    className={editorTreeGroupClassName}
                                    onClick={(e) => handleGroupClick(e, group)}
                                    onContextMenu={(e) =>
                                        handleHeaderRightClick(e, group)
                                    }
                                    key={index}
                                >
                                    {localize(
                                        'sidebar.explore.openEditor.group',
                                        'Group',
                                        (index + 1).toString()
                                    )}
                                    {groupToolbar && (
                                        <Toolbar
                                            data={groupToolbar}
                                            onClick={(e, item) =>
                                                handleToolBarClick(
                                                    e,
                                                    item,
                                                    group
                                                )
                                            }
                                        />
                                    )}
                                </div>
                            )}
                            {group.data?.map((file) => {
                                const isActive =
                                    group.id === current?.id &&
                                    file.id === current?.tab?.id;
                                return (
                                    <div
                                        title={
                                            file.data?.path &&
                                            `${file.data?.path}/${file.name}`
                                        }
                                        className={classNames(
                                            editorTreeItemClassName,
                                            isActive &&
                                                editorTreeActiveItemClassName
                                        )}
                                        tabIndex={0}
                                        key={`${index}_${file.id}`}
                                        onClick={() =>
                                            handleItemClick(group, file)
                                        }
                                        onContextMenu={(e) =>
                                            handleRightClick(e, group, file)
                                        }
                                    >
                                        <Icon
                                            className={
                                                editorTreeCloseIconClassName
                                            }
                                            onClick={() =>
                                                handleCloseClick(group, file)
                                            }
                                            type="close"
                                        />
                                        <Icon
                                            className={
                                                editorTreeFileIconClassName
                                            }
                                            type={file.data?.icon || file.icon}
                                        />
                                        <span
                                            className={
                                                editorTreeFileNameClassName
                                            }
                                        >
                                            {file.name}
                                        </span>
                                        <span
                                            className={
                                                editorTreeFilePathClassName
                                            }
                                        >
                                            {file.data?.path}
                                        </span>
                                    </div>
                                );
                            })}
                        </React.Fragment>
                    );
                })}
            </div>
        </Scrollable>
    );
};

export { EditorTree };
