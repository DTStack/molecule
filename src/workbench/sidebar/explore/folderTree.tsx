import 'reflect-metadata';
import React, { memo, useRef, useEffect, useLayoutEffect } from 'react';
import { IFolderTree, IFolderTreeNodeProps } from 'mo/model';
import { select, getEventPosition } from 'mo/common/dom';
import Tree from 'mo/components/tree';
import { IMenuItemProps, Menu } from 'mo/components/menu';
import { Button } from 'mo/components/button';
import type { IFolderTreeController } from 'mo/controller/explorer/folderTree';
import { useContextView } from 'mo/components/contextView';
import { useContextMenu } from 'mo/components/contextMenu';
import {
    folderTreeClassName,
    folderTreeEditClassName,
    folderTreeInputClassName,
} from './base';
import { classNames } from 'mo/common/className';
import { Scrollable } from 'mo/components';
import { ICollapseItem } from 'mo/components/collapse';

export interface IFolderTreeProps extends IFolderTreeController, IFolderTree {
    panel: ICollapseItem;
}

const detectHasEditableStatus = (data) => {
    const stack = [...data];
    let res = false;
    while (stack.length) {
        const headElm = stack.pop();
        if (headElm?.isEditable) {
            res = true;
            break;
        } else {
            stack.push(...(headElm?.children || []));
        }
    }
    return res;
};

/**
 * A simple wrapper Input, achieve autoFucus & auto select file name
 */
const Input = React.forwardRef(
    (
        // same as raw input
        props: React.DetailedHTMLProps<
            React.InputHTMLAttributes<HTMLInputElement>,
            HTMLInputElement
        >,
        ref: React.LegacyRef<HTMLInputElement>
    ) => {
        const inputRef = useRef<HTMLInputElement>(null);
        useLayoutEffect(() => {
            if (inputRef.current) {
                const ext = ((props.defaultValue || '') as string).lastIndexOf(
                    '.'
                );
                inputRef.current.selectionStart = 0;
                inputRef.current.selectionEnd =
                    // if period at position of 0, then this period means hidden file
                    ext > 0
                        ? ext
                        : ((props.defaultValue || '') as string).length;
            }
        }, []);
        return <input {...props} ref={inputRef} />;
    }
);

const FolderTree: React.FunctionComponent<IFolderTreeProps> = (props) => {
    const {
        folderTree = {},
        entry,
        panel,
        onUpdateFileName,
        onSelectFile,
        onDropTree,
        onClickContextMenu,
        onRightClick,
        onLoadData,
        createTreeNode,
        onExpandKeys,
        ...restProps
    } = props;

    const {
        data = [],
        folderPanelContextMenu = [],
        expandKeys,
        loadedKeys,
        current,
    } = folderTree;

    const inputRef = useRef<HTMLInputElement>(null);
    // tree context view
    const contextMenu = useRef<ReturnType<typeof useContextMenu>>();

    // panel context view
    const contextView = useContextView();

    // to detect current tree whether is editable
    const hasEditable = detectHasEditableStatus(data);

    const onClickMenuItem = (e, item) => {
        onClickContextMenu?.(item);
        contextMenu.current?.hide();
    };

    // init context menu
    const initContextMenu = () => {
        return useContextMenu({
            anchor: select(`.${folderTreeClassName}`),
            render: () => (
                <Menu
                    role="menu"
                    onClick={onClickMenuItem}
                    data={folderPanelContextMenu}
                />
            ),
        });
    };

    const handleMenuClick = (
        item: IMenuItemProps,
        data: IFolderTreeNodeProps
    ) => {
        onClickContextMenu?.(item, data);
        contextView.hide();
    };

    const handleRightClick = (event, data) => {
        const menuItems = onRightClick?.(data) || [];

        menuItems.length &&
            contextView?.show(getEventPosition(event), () => (
                <Menu
                    role="menu"
                    onClick={(_, item) => handleMenuClick(item!, data)}
                    data={menuItems}
                />
            ));
    };

    const handleUpdateFile = (
        e: HTMLInputElement,
        node: IFolderTreeNodeProps
    ) => {
        const newName = e.value;
        onUpdateFileName?.({
            ...node,
            name: newName,
        });
    };

    /**
     * update file info when input blur
     */
    const handleInputBlur = (
        e: React.FocusEvent<HTMLInputElement>,
        node: IFolderTreeNodeProps
    ) => {
        handleUpdateFile(e.target, node);
    };

    /**
     * update file info when press `Enter` or `esc`
     */
    const handleInputKeyDown = (
        e: React.KeyboardEvent<HTMLInputElement>,
        node: IFolderTreeNodeProps
    ) => {
        if (e.keyCode === 13 || e.keyCode === 27) {
            handleUpdateFile(e.target as EventTarget & HTMLInputElement, node);
        }
    };

    const renderTitle = (node: IFolderTreeNodeProps) => {
        const { isEditable, name } = node;

        return isEditable ? (
            <Input
                role="input"
                className={folderTreeInputClassName}
                type="text"
                defaultValue={name}
                ref={inputRef}
                onKeyDown={(e) => handleInputKeyDown(e, node)}
                autoComplete="off"
                autoFocus
                onBlur={(e) => handleInputBlur(e, node)}
                onClick={(e) => e.stopPropagation()}
            />
        ) : (
            name!
        );
    };

    const handleTreeClick = () => {
        onSelectFile?.();
    };

    const handleDropTree = (source, target) => {
        onDropTree?.(source, target);
    };

    const handleAddRootFolder = () => {
        createTreeNode?.('RootFolder');
    };

    useEffect(() => {
        if (folderPanelContextMenu.length > 0) {
            contextMenu.current = initContextMenu();
        }
        return () => {
            contextMenu.current?.dispose();
        };
    }, [data.length]);

    const welcomePage = (
        <div data-content={panel.id}>
            {entry ? (
                <>{entry}</>
            ) : (
                <div style={{ padding: '10px 5px' }}>
                    you have not yet opened a folder
                    <Button onClick={handleAddRootFolder}>Add Folder</Button>
                </div>
            )}
        </div>
    );

    if (!data.length) return welcomePage;

    return (
        <Scrollable noScrollX isShowShadow>
            <div data-content={panel.id} style={{ height: '100%' }}>
                <Tree
                    // root folder do not render
                    activeKey={current?.id}
                    expandKeys={expandKeys}
                    loadedKeys={loadedKeys}
                    data={data[0]?.children || []}
                    className={classNames(
                        folderTreeClassName,
                        hasEditable && folderTreeEditClassName
                    )}
                    draggable={!hasEditable}
                    onDropTree={handleDropTree}
                    onSelect={onSelectFile}
                    onTreeClick={handleTreeClick}
                    onRightClick={handleRightClick}
                    renderTitle={renderTitle}
                    onLoadData={onLoadData}
                    onExpand={onExpandKeys}
                    {...restProps}
                />
            </div>
        </Scrollable>
    );
};
export default memo(FolderTree);
