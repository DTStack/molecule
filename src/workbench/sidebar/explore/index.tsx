import 'reflect-metadata';
import { connect } from 'mo/react';
import { container } from 'tsyringe';
import { FolderTreeService } from 'mo/services';
import { Explorer } from './explore';
import FolderTree from './folderTree';
import { FolderTreeController } from 'mo/controller/explorer/folderTree';
import { EditorTree } from './editorTree';

const folderTreeService = container.resolve(FolderTreeService);
const folderTreeController = container.resolve(FolderTreeController);

const FolderTreeView = connect(
    folderTreeService,
    FolderTree,
    folderTreeController
);

export { Explorer, FolderTreeView, FolderTree, EditorTree };
