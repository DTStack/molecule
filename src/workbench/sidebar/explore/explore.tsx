import React from 'react';
import { Collapse } from 'mo/components/collapse';
import { Header, Content } from 'mo/workbench/sidebar';
import { IExplorer } from 'mo/model/workbench/explorer/explorer';
import { IExplorerController } from 'mo/controller/explorer/explorer';
import { Toolbar } from 'mo/components/toolbar';
import { defaultExplorerClassName } from './base';
import { localize } from 'mo/i18n/localize';

type IExplorerProps = IExplorer & IExplorerController;

export const Explorer: React.FunctionComponent<IExplorerProps> = (
    props: IExplorerProps
) => {
    const {
        activePanelKeys,
        data = [],
        headerToolBar,
        onClick,
        onActionsContextMenuClick,
        onCollapseChange,
        onToolbarClick,
    } = props;
    return (
        <div className={defaultExplorerClassName}>
            <Header
                title={localize('sidebar.explore.title', 'Explorer')}
                toolbar={
                    <Toolbar
                        data={[headerToolBar!]}
                        onClick={onClick}
                        onContextMenuClick={onActionsContextMenuClick}
                    />
                }
            />
            <Content>
                <Collapse
                    data={data}
                    activePanelKeys={activePanelKeys}
                    onCollapseChange={onCollapseChange}
                    onToolbarClick={onToolbarClick}
                />
            </Content>
        </div>
    );
};
