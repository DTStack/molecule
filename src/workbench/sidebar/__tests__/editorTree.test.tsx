import 'reflect-metadata';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { cleanup, fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { expectFnCalled } from '@test/utils';

import { EditorTree, IOpenEditProps } from '../explore/editorTree';
import {
    editorTreeActiveItemClassName,
    editorTreeGroupClassName,
} from '../explore/base';
import { constants } from 'mo/services/builtinService/const';
import { container } from 'tsyringe';
import { LocaleService } from 'mo/i18n';
import { ExtendsLocales } from 'mo/extensions/locales-defaults';

const PaneEditorTree = (props: Omit<IOpenEditProps, 'panel'>) => {
    return <EditorTree panel={{ id: 'test', name: 'test' }} {...props} />;
};

const mockTab1 = {
    id: 'tab1',
    name: 'tab1',
    data: {
        path: 'tab',
    },
};

const mockTab2 = {
    id: 'tab2',
    name: 'tab2',
    data: {
        path: 'tab',
    },
};
const TAB1_PATH = `${mockTab1.data.path}/${mockTab1.name}`;
const TAB2_PATH = `${mockTab2.data.path}/${mockTab2.name}`;

const mockGroups = [
    {
        id: 1,
        tab: {
            id: 'tab1',
            name: 'tab1',
        },
        data: [mockTab1, mockTab2],
    },
];

// mock Toolbar component
jest.mock('mo/components/toolbar', () => {
    const originalModule = jest.requireActual('mo/components/toolbar');
    return {
        ...originalModule,
        Toolbar: ({ onClick, data = [] }) => (
            <div data-testid="toolbar">
                {data.map((item, index) => (
                    <span key={index} onClick={(e) => onClick(e, item)}>
                        toolbar-{index}
                    </span>
                ))}
            </div>
        ),
    };
});

// to mock useRef
jest.mock('react', () => {
    const originReact = jest.requireActual('react');
    return {
        ...originReact,
        useRef: jest.fn(() => ({
            current: null,
        })),
    };
});

describe('The EditorTree Component', () => {
    // initial locales
    const localeService = container.resolve(LocaleService);
    localeService.initialize(ExtendsLocales.contributes!.languages!, 'en');

    afterEach(cleanup);

    test('Match Snapshot', () => {
        let component;
        act(() => {
            component = renderer.create(<PaneEditorTree groups={mockGroups} />);
        });
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Should get null without groups', () => {
        const { container } = render(<PaneEditorTree />);
        expect(container.innerHTML).toBe('');
    });

    test('Should not render title without path in data', () => {
        const { container } = render(
            <PaneEditorTree
                groups={[
                    {
                        id: 1,
                        tab: {
                            id: 'tab1',
                            name: 'tab1',
                        },
                        data: [
                            {
                                id: 'tab1',
                                name: 'tab1',
                            },
                            {
                                id: 'tab2',
                                name: 'tab2',
                            },
                        ],
                    },
                ]}
            />
        );
        expect(container.querySelector("*[title='tab/tab1']")).toBeNull();
    });

    test('Should support to active the pane', () => {
        const { getByTitle } = render(
            <PaneEditorTree groups={mockGroups} current={mockGroups[0]} />
        );

        const tab1 = getByTitle('tab/tab1');
        expect(tab1.classList).toContain(editorTreeActiveItemClassName);
    });

    test('Should support to close tab', () => {
        expectFnCalled((testFn) => {
            const { getByTitle } = render(
                <PaneEditorTree
                    groups={mockGroups}
                    onClose={testFn}
                    current={mockGroups[0]}
                />
            );

            const tab1 = getByTitle(TAB1_PATH);
            fireEvent.click(tab1.querySelector('span.codicon-close')!);

            expect(testFn.mock.calls[0][0]).toBe(mockTab1.name);
            expect(testFn.mock.calls[0][1]).toBe(1);
        });
    });

    test('Should support to contextMenu event', () => {
        const contextMenu = {
            id: 'test',
            name: 'menu',
        };
        expectFnCalled((contextMenuFn) => {
            const { getByTitle, getByRole } = render(
                <PaneEditorTree
                    groups={mockGroups}
                    current={mockGroups[0]}
                    contextMenu={[contextMenu]}
                    onContextMenu={contextMenuFn}
                />
            );

            const tab1 = getByTitle(TAB1_PATH);
            fireEvent.contextMenu(tab1);

            const menu = getByRole('menu');
            expect(menu).toBeInTheDocument();

            fireEvent.click(menu.firstElementChild!);

            expect(contextMenuFn.mock.calls[0][0]).toEqual(
                expect.objectContaining(contextMenu)
            );
            expect(contextMenuFn.mock.calls[0][1]).toEqual(1);
            expect(contextMenuFn.mock.calls[0][2]).toEqual(mockTab1);
        });
    });

    test('Should support to select tab', () => {
        expectFnCalled((selectFn) => {
            const { getByTitle } = render(
                <PaneEditorTree
                    groups={mockGroups}
                    current={mockGroups[0]}
                    onSelect={selectFn}
                />
            );

            const tab1 = getByTitle(TAB1_PATH);
            fireEvent.click(tab1);
            // same current tab won't triiger onSelect event
            expect(selectFn).not.toBeCalled();

            const tab2 = getByTitle(TAB2_PATH);
            fireEvent.click(tab2);
        });
    });

    test('Should not to scroll into view when scollerHeight is invalid', () => {
        const mockScroll = jest.fn();
        (React.useRef as jest.Mock).mockImplementation(() => ({
            current: { scrollHeight: 100, scrollTo: mockScroll },
        }));

        const current = mockGroups[0];

        render(<PaneEditorTree groups={mockGroups} current={current} />);
        expect(mockScroll).not.toBeCalled();
    });

    test('Should scroll into view when add an active tab', () => {
        const mockScroll = jest.fn();
        (React.useRef as jest.Mock).mockImplementation(() => ({
            current: { scrollHeight: 200, scrollTo: mockScroll },
        }));

        const current = mockGroups[0];

        render(<PaneEditorTree groups={mockGroups} current={current} />);
        expect(mockScroll).toBeCalled();
    });
});

const multipleGroups = [
    {
        id: 1,
        tab: {
            id: 'tab1',
            name: 'tab1',
        },
        data: [mockTab1, mockTab2],
    },
    {
        id: 2,
        tab: {
            id: 'tab1',
            name: 'tab1',
        },
        data: [mockTab1, mockTab2],
    },
];

describe('The EditorTree Component With multiple groups', () => {
    afterEach(cleanup);

    test('Match Snapshot', () => {
        const component = renderer.create(
            <PaneEditorTree groups={multipleGroups} />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Should focus the first child in group when click group title', () => {
        expectFnCalled((selectFn) => {
            const { container } = render(
                <PaneEditorTree
                    groups={multipleGroups}
                    current={multipleGroups[0]}
                    onSelect={selectFn}
                />
            );

            const groups = container.querySelectorAll(
                `.${editorTreeGroupClassName}`
            );

            expect(groups).toHaveLength(2);

            const group1 = groups[0];
            fireEvent.click(group1);

            expect(
                (group1.nextElementSibling! as HTMLDivElement).focus
            ).toBeTruthy();

            expect(selectFn.mock.calls[0][0]).toBe(mockTab1.name);
            expect(selectFn.mock.calls[0][1]).toBe(multipleGroups[0].id);
        });
    });

    test('Should support trigger contextMenu event in group title', () => {
        const contextMenu = {
            id: 'test',
            name: 'menu',
        };
        expectFnCalled((contextMenuFn) => {
            const { container, getByRole } = render(
                <PaneEditorTree
                    groups={multipleGroups}
                    current={multipleGroups[0]}
                    headerContextMenu={[contextMenu]}
                    onContextMenu={contextMenuFn}
                />
            );

            const group1 = container.querySelector(
                `.${editorTreeGroupClassName}`
            );

            fireEvent.contextMenu(group1!);

            const menu = getByRole('menu');
            expect(menu).toBeInTheDocument();

            fireEvent.click(menu.firstElementChild!);

            expect(contextMenuFn.mock.calls[0][0]).toEqual(
                expect.objectContaining(contextMenu)
            );
            expect(contextMenuFn.mock.calls[0][1]).toEqual(
                multipleGroups[0].id
            );
            expect(contextMenuFn.mock.calls[0][2]).toBeUndefined();
        });
    });

    test('Should have a toolbar in group title', () => {
        const closeMockFn = jest.fn();
        const saveMockFn = jest.fn();
        const toolbarMockFn = jest.fn();

        const { getAllByTestId } = render(
            <PaneEditorTree
                groups={multipleGroups}
                current={multipleGroups[0]}
                groupToolbar={[
                    {
                        id: constants.EXPLORER_TOGGLE_SAVE_GROUP,
                        title: 'Save Group',
                        icon: 'save-all',
                    },
                    {
                        id: constants.EXPLORER_TOGGLE_CLOSE_GROUP_EDITORS,
                        title: 'Close Group Editors',
                        icon: 'close-all',
                    },
                    {
                        id: 'test',
                        title: 'testing',
                    },
                ]}
                onCloseGroup={closeMockFn}
                onSaveGroup={saveMockFn}
                onToolbarClick={toolbarMockFn}
            />
        );

        const toolbars = getAllByTestId('toolbar');

        expect(toolbars).toHaveLength(multipleGroups.length);

        const index = 0;
        let triggerBtn = toolbars[index].firstElementChild!;

        fireEvent.click(triggerBtn);

        // first button is for saving
        expect(saveMockFn).toBeCalled();
        expect(saveMockFn.mock.calls[0][0]).toBe(multipleGroups[index].id);

        triggerBtn = triggerBtn.nextElementSibling!;
        fireEvent.click(triggerBtn);

        // next button is for closing
        expect(closeMockFn).toBeCalled();
        expect(closeMockFn.mock.calls[0][0]).toBe(multipleGroups[index].id);

        triggerBtn = triggerBtn.nextElementSibling!;
        fireEvent.click(triggerBtn);

        // last button is extra button
        expect(toolbarMockFn).toBeCalled();
        expect(toolbarMockFn.mock.calls[0][0]).toEqual({
            id: 'test',
            title: 'testing',
        });
        expect(toolbarMockFn.mock.calls[0][1]).toBe(multipleGroups[index].id);
    });
});
