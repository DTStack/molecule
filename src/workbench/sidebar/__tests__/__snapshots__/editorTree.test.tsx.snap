// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The EditorTree Component Match Snapshot 1`] = `
<div
  className="mo-editorTree"
  data-content="test"
>
  <div
    className="mo-editorTree__item"
    onClick={[Function]}
    onContextMenu={[Function]}
    tabIndex={0}
    title="tab/tab1"
  >
    <span
      className="mo-editorTree__close codicon codicon-close"
      onClick={[Function]}
    />
    <span
      className="mo-editorTree__item__fileName"
    >
      tab1
    </span>
    <span
      className="mo-editorTree__item__filePath"
    >
      tab
    </span>
  </div>
  <div
    className="mo-editorTree__item"
    onClick={[Function]}
    onContextMenu={[Function]}
    tabIndex={0}
    title="tab/tab2"
  >
    <span
      className="mo-editorTree__close codicon codicon-close"
      onClick={[Function]}
    />
    <span
      className="mo-editorTree__item__fileName"
    >
      tab2
    </span>
    <span
      className="mo-editorTree__item__filePath"
    >
      tab
    </span>
  </div>
</div>
`;

exports[`The EditorTree Component With multiple groups Match Snapshot 1`] = `
<div
  className="mo-editorTree"
  data-content="test"
>
  <div
    className="mo-editorTree__group"
    onClick={[Function]}
    onContextMenu={[Function]}
  >
    第 1 组
  </div>
  <div
    className="mo-editorTree__item"
    onClick={[Function]}
    onContextMenu={[Function]}
    tabIndex={0}
    title="tab/tab1"
  >
    <span
      className="mo-editorTree__close codicon codicon-close"
      onClick={[Function]}
    />
    <span
      className="mo-editorTree__item__fileName"
    >
      tab1
    </span>
    <span
      className="mo-editorTree__item__filePath"
    >
      tab
    </span>
  </div>
  <div
    className="mo-editorTree__item"
    onClick={[Function]}
    onContextMenu={[Function]}
    tabIndex={0}
    title="tab/tab2"
  >
    <span
      className="mo-editorTree__close codicon codicon-close"
      onClick={[Function]}
    />
    <span
      className="mo-editorTree__item__fileName"
    >
      tab2
    </span>
    <span
      className="mo-editorTree__item__filePath"
    >
      tab
    </span>
  </div>
  <div
    className="mo-editorTree__group"
    onClick={[Function]}
    onContextMenu={[Function]}
  >
    第 2 组
  </div>
  <div
    className="mo-editorTree__item"
    onClick={[Function]}
    onContextMenu={[Function]}
    tabIndex={0}
    title="tab/tab1"
  >
    <span
      className="mo-editorTree__close codicon codicon-close"
      onClick={[Function]}
    />
    <span
      className="mo-editorTree__item__fileName"
    >
      tab1
    </span>
    <span
      className="mo-editorTree__item__filePath"
    >
      tab
    </span>
  </div>
  <div
    className="mo-editorTree__item"
    onClick={[Function]}
    onContextMenu={[Function]}
    tabIndex={0}
    title="tab/tab2"
  >
    <span
      className="mo-editorTree__close codicon codicon-close"
      onClick={[Function]}
    />
    <span
      className="mo-editorTree__item__fileName"
    >
      tab2
    </span>
    <span
      className="mo-editorTree__item__filePath"
    >
      tab
    </span>
  </div>
</div>
`;
